### PR TITLE
refactor: use config crate for service

### LINKED ISSUE #402: Use `config` crate for service

### PR DESCRIPTION
**Pull request by shiyasmohd**
This PR closes #402 

Changes in this PR
- Updated imports in `indexer_service.rs` & `service/src/service.rs`
- Delete `config.rs` from `common/indexer-service` & `service`
- Added a fn to get as `SocketAddr` from `MetricsConfig`

### CODE REVIEW

File: service/src/service.rs [hunk 1]
```diff
@@ -129,7 +129,7 @@ pub async fn run() -> anyhow::Result<()> {
     // Load the json-rpc service configuration, which is a combination of the
     // general configuration options for any indexer service and specific
     // options added for JSON-RPC
-    let config = MainConfig::parse(indexer_config::ConfigPrefix::Service, cli.config.as_ref())
```
Comment (gusinacio) [inline service/src/service.rs:132]: Hey, I think you can `Box::leak` the configuration, that way you don't need to clone things.

File: service/src/service.rs [hunk 2]
```diff
@@ -151,32 +149,21 @@ pub async fn run() -> anyhow::Result<()> {
...
-            .clone(),
-        graph_node_query_base_url: config
-            .0
-            .graph_node
-            .as_ref()
-            .expect("config must have `common.graph_node.query_url` set")
-            .query_base_url
-            .clone(),
+        graph_node_status_url: config.graph_node.status_url.clone().to_string(),
+        graph_node_query_base_url: config.graph_node.query_url.clone().to_string(),
```
Comment (gusinacio) [inline service/src/service.rs:161]: You could convert this into `Url`

File: common/src/indexer_service/http/indexer_service.rs [hunk 3]
```diff
@@ -233,50 +233,67 @@ impl IndexerService {
...
             Duration::from_secs(
                 options
                     .config
-                    .network_subgraph
-                    .recently_closed_allocation_buffer_seconds,
+                    .subgraphs
+                    .network
+                    .config
+                    .syncing_interval_secs
+                    .as_secs(),
```
Comment (gusinacio) [inline common/src/indexer_service/http/indexer_service.rs:243]: syncing_interval is already a `Duration`

File: common/src/indexer_service/http/indexer_service.rs [hunk 4]
```diff
@@ -233,50 +233,67 @@ impl IndexerService {
...
+            ),
+            Duration::from_secs(
+                options
+                    .config
+                    .subgraphs
+                    .network
+                    .recently_closed_allocation_buffer_secs
+                    .as_secs(),
```
Comment (gusinacio) [inline common/src/indexer_service/http/indexer_service.rs:251]: recently_closed_allocatin_buffer_secs is already a `Duration`

File: common/src/indexer_service/http/indexer_service.rs [hunk 5]
```diff
@@ -233,50 +233,67 @@ impl IndexerService {
...
 
         // Maintain an up-to-date set of attestation signers, one for each
         // allocation
         let attestation_signers = attestation_signers(
             allocations.clone(),
-            options.config.indexer.operator_mnemonic.clone(),
-            options.config.graph_network.chain_id,
+            options.config.indexer.operator_mnemonic.to_string(),
+            options.config.blockchain.chain_id.clone() as u64,
             dispute_manager,
```
Comment (gusinacio) [inline common/src/indexer_service/http/indexer_service.rs:261]: You could update `attestation_signers` to receive a `Mnemonic`

File: common/src/indexer_service/http/indexer_service.rs [hunk 6]
```diff
@@ -233,50 +233,67 @@ impl IndexerService {
...
-                        deployment,
-                    )
-                })
-                .transpose()?,
+            Some(DeploymentDetails {
+                deployment: options.config.subgraphs.escrow.config.deployment_id,
+                status_url: Some(options.config.graph_node.status_url.clone()),
+                query_url: options.config.graph_node.query_url.clone(),
+                query_auth_token: options.config.service.free_query_auth_token.clone(),
+            }),
```
Comment (gusinacio) [inline common/src/indexer_service/http/indexer_service.rs:272]: wonderful

File: common/src/indexer_service/http/indexer_service.rs [hunk 7]
```diff
@@ -233,50 +233,67 @@ impl IndexerService {
...
-            Duration::from_secs(options.config.escrow_subgraph.syncing_interval),
+            Duration::from_secs(
+                options
+                    .config
+                    .subgraphs
+                    .escrow
+                    .config
+                    .syncing_interval_secs
+                    .as_secs(),
+            ),
```
Comment (gusinacio) [inline common/src/indexer_service/http/indexer_service.rs:296]: syncing_interval is already a `Duration`

File: common/src/indexer_service/http/indexer_service.rs [hunk 8]
```diff
@@ -290,19 +307,32 @@ impl IndexerService {
...
-        let timestamp_error_tolerance =
-            Duration::from_secs(options.config.tap.timestamp_error_tolerance);
+        let timestamp_error_tolerance = Duration::from_secs(
+            options
+                .config
+                .tap
+                .rav_request
+                .timestamp_buffer_secs
+                .as_secs(),
+        );
```
Comment (gusinacio) [inline common/src/indexer_service/http/indexer_service.rs:334]: Already `Duration`.

File: common/src/indexer_service/http/indexer_service.rs [hunk 9]
```diff
@@ -343,7 +373,7 @@ impl IndexerService {
         };
 
         let operator_address = Json(
-            serde_json::json!({ "publicKey": public_key(&options.config.indexer.operator_mnemonic)?}),
+            serde_json::json!({ "publicKey": public_key(&options.config.indexer.operator_mnemonic.to_string())?}),
```
Comment (gusinacio) [inline common/src/indexer_service/http/indexer_service.rs:376]: Maybe update `public_key` to receive Mnemonic

File: config/src/config.rs [hunk 10]
```diff
@@ -251,28 +257,34 @@ impl DatabaseConfig {
...
 pub struct MetricsConfig {
     pub port: u16,
 }
 
-#[derive(Debug, Deserialize)]
+impl MetricsConfig {
+    pub fn get_socket_addr(&self) -> SocketAddr {
+        SocketAddr::V4(SocketAddrV4::new(Ipv4Addr::new(0, 0, 0, 0), self.port))
+    }
+}
```
Comment (gusinacio) [inline config/src/config.rs:277]: nice

**Review by gusinacio — CHANGES_REQUESTED** [review]
This is great code! 

In general, I'd like to remove those conversions and use proper types til the last minute. 
`indexer_config` already does a lot of validation and conversion while parsing the config to give you the correct types, let's not convert multiple times.

**Comment by shiyasmohd** [comment]
Hey @gusinacio ,
Thanks for the feedback. I've made necessary changes.
I've also refactored some lines in [this](https://github.com/graphprotocol/indexer-rs/pull/409/commits/c0358bb24d3c238356fdfc7804848880ea32c440#diff-2173e0a1ff7c183bf7d8e97977ffddd00ac5e3ccc45b8d1d3c59723c90fa258c) commit. 

Please let me know your thoughts.

**Review by gusinacio — DISMISSED** [review]
LGTM, thank you for your beautiful code!

**Review by gusinacio — APPROVED** [review]

**Pull request merged**